tests/pubsub/testgossipsub.nim

```diff
@@ -220,6 +220,70 @@ suite "GossipSub":
 
     await allFuturesThrowing(nodesFut.concat())
 
+  asyncTest "GossipSub's observers should run after message is sent, received and validated":
+    var
+      handlerFut = newFuture[bool]()
+      recvCounter = 0
+      sendCounter = 0
+      validatedCounter = 0
+
+    proc handler(topic: string, data: seq[byte]) {.async.} =
+      check topic == "foo"
+      handlerFut.complete(true)
+
+    proc onRecv(peer: PubSubPeer, msgs: var RPCMsg) =
+      inc recvCounter
+
+    proc onSend(peer: PubSubPeer, msgs: var RPCMsg) =
+      inc sendCounter
+
+    proc onValidated(peer: PubSubPeer, msg: Message, msgId: MessageId) =
+      inc validatedCounter
+
+    let obs0 = PubSubObserver(onSend: onSend)
+    let obs1 = PubSubObserver(onRecv: onRecv, onValidated: onValidated)
+
+    let
+      nodes = generateNodes(2, gossip = true)
+      # start switches
+      nodesFut = await allFinished(nodes[0].switch.start(), nodes[1].switch.start())
```

Comment:
is nodesFut necessary?

Reply:
It is for the last line, for `await allFuturesThrowing(nodesFut.concat())`, to raise an exception in case one switch failed starting.

Basically all tests in that file do it this way, so just followed the existing pattern. I would personally do the check right after attempting to start the switches and not at the end of the test, but wanted to be consistent with all the other tests. And maybe I was missing something.

LMK how you prefer

Reply:
I don't believe this is necessary. Don't always take the existing code as a good solution.

Reply:
Yes for sure, but for someone who is not very familiar with the codebase, sometimes the safest and most natural approach is to follow the patterns done by the people who wrote it :)

Reply:
Removed `nodesFut`, please take a look now

Reply:
Agreed, it was just an observation ;)